ff_ext/src/babybear.rs

```diff
@@ -0,0 +1,184 @@
+        0x3c318939, 0x0b5b2f22, 0x2edb11d5, 0x213effdf, 0x0cac4606, 0x241af16d,
+    ]);
+
+    impl FieldFrom<u64> for BabyBear {
+        fn from_v(v: u64) -> Self {
+            Self::from_u64(v)
+        }
+    }
+
+    impl FieldFrom<u64> for BabyBearExt4 {
+        fn from_v(v: u64) -> Self {
+            Self::from_u64(v)
+        }
+    }
+
+    impl FieldInto<BabyBear> for BabyBear {
+        fn into_f(self) -> BabyBear {
+            self
+        }
+    }
+
+    impl PoseidonField for BabyBear {
+        type P = Poseidon2BabyBear<POSEIDON2_BABYBEAR_WIDTH>;
+        type T = DuplexChallenger<Self, Self::P, POSEIDON2_BABYBEAR_WIDTH, POSEIDON2_BABYBEAR_RATE>;
+        fn get_default_challenger() -> Self::T {
+            let p = Poseidon2BabyBear::new(
+                ExternalLayerConstants::new(
+                    BABYBEAR_RC16_EXTERNAL_INITIAL.to_vec(),
+                    BABYBEAR_RC16_EXTERNAL_FINAL.to_vec(),
+                ),
+                BABYBEAR_RC16_INTERNAL.to_vec(),
+            );
+            DuplexChallenger::<
+                Self,
+                Self::P,
+                POSEIDON2_BABYBEAR_WIDTH,
+                POSEIDON2_BABYBEAR_RATE,
+            >::new(p)
+        }
+    }
+
+    impl FromUniformBytes for BabyBear {
+        type Bytes = [u8; 8];
+
+        fn try_from_uniform_bytes(bytes: [u8; 8]) -> Option<Self> {
+            let value = u32::from_le_bytes(bytes[..4].try_into().unwrap());
+            let is_canonical = value < Self::ORDER_U32;
+            is_canonical.then(|| Self::from_u32(value))
+        }
+    }
+
+    impl SmallField for BabyBear {
+        const MODULUS_U64: u64 = Self::ORDER_U32 as u64;
+
+        /// Convert a byte string into a list of field elements
+        fn bytes_to_field_elements(bytes: &[u8]) -> Vec<Self> {
+            bytes
+                .chunks(8)
+                .map(|chunk| {
+                    let mut array = [0u8; 8];
+                    array[..chunk.len()].copy_from_slice(chunk);
+                    unsafe { std::ptr::read_unaligned(array.as_ptr() as *const u64) }
+                })
+                .map(Self::from_u64)
+                .collect::<Vec<_>>()
+        }
+
+        /// Convert a field elements to a u64.
+        fn to_canonical_u64(&self) -> u64 {
+            self.as_canonical_u32() as u64
+        }
+    }
+
+    impl_from_uniform_bytes_for_binomial_extension!(p3::babybear::BabyBear, 4);
+
+    impl ExtensionField for BabyBearExt4 {
+        const DEGREE: usize = 4;
+        const MULTIPLICATIVE_GENERATOR: Self = <BabyBearExt4 as Field>::GENERATOR;
+        const TWO_ADICITY: usize = BabyBear::TWO_ADICITY;
+        // Passing two-adacity itself to this function will get the root of unity
+        // with the largest order, i.e., order = 2^two-adacity.
+        const BASE_TWO_ADIC_ROOT_OF_UNITY: Self::BaseField = BabyBear::new(0x78000000);
+        const TWO_ADIC_ROOT_OF_UNITY: Self =
```

Comment:
we can remove this after  #858 #843, as this is only for field wrapper to serve ark lib

Reply:
This is removed in #966.